Fix installing in 'venv' and datetime tests on Windows

lycantropos opened the pull request:
Resolves #468.

- [x] Run `cargo fmt` (This is checked by travis ci).
- [x] Run `cargo clippy` and check there are no hard errors (There are a bunch of existing warnings; This is also checked by travis).
- [x] If applicable, add an entry in the changelog.
- [x] If applicable, add documentation to all new items and extend the guide.
- [x] If applicable, add tests for all new or fixed functions

lycantropos commented on tests/test_venv.rs:
```diff
@@ -0,0 +1,61 @@
+                    format!(
+                        "{0}\\Scripts\\activate.bat && \
+                         cd examples\\{1} && \
+                         python -m pip install --upgrade setuptools-rust && \
```
without pre-installing `setuptools-rust` package command
```cmd
python setup.py build
```
fails with
```cmd
running build_ext
running build_rust
error: invalid command 'build_rust'
```
so it should be installed before

pganssle replied:
@lycantropos Actually you should not invoke `setup.py` directly, instead use `pip install` with a `pyproject.toml`. The `word-count` and `rustapi_modules` examples are built with PEP 517, and installing or building them by invoking `setup.py` probably works even worse than it would in a normal project.

lycantropos replied:
```cmd
(venv) > python -m pip install --upgrade pip
Requirement already up-to-date: pip in c:\users\azati\clionprojects\pyo3\venv\lib\site-packages (19.1)
(venv) > python -m pip install .
```
gives
```cmd
> python -m pip install .
Processing c:\users\azati\clionprojects\pyo3\examples\word-count
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
Building wheels for collected packages: word-count
  Building wheel for word-count (PEP 517) ... error
  ERROR: Complete output from command 'C:\Users\azati\CLionProjects\pyo3\venv\Scripts\python.exe' 'C:\Users\azati\CLionProjects\pyo3\venv\lib\site-packages\pip\_vendor\pep51
7\_in_process.py' build_wheel 'C:\Users\azati\AppData\Local\Temp\tmp1rcvfwh_':
  ERROR: running bdist_wheel
  running build
  running build_py
  running egg_info
  writing word_count.egg-info\PKG-INFO
  writing top-level names to word_count.egg-info\top_level.txt
  writing dependency_links to word_count.egg-info\dependency_links.txt
  reading manifest file 'word_count.egg-info\SOURCES.txt'
  reading manifest template 'MANIFEST.in'
  writing manifest file 'word_count.egg-info\SOURCES.txt'
  running build_ext
  running build_rust
  error: failed to load source for a dependency on `pyo3`

  Caused by:
    Unable to update C:\Users\azati\AppData\Local

  Caused by:
    failed to read `C:\Users\azati\AppData\Local\Cargo.toml`

  Caused by:
    The system cannot find the file specified. (os error 2)
  Traceback (most recent call last):
    File "C:\Users\azati\CLionProjects\pyo3\venv\lib\site-packages\pip\_vendor\pep517\_in_process.py", line 207, in <module>
      main()
    File "C:\Users\azati\CLionProjects\pyo3\venv\lib\site-packages\pip\_vendor\pep517\_in_process.py", line 197, in main
      json_out['return_val'] = hook(**hook_input['kwargs'])
    File "C:\Users\azati\CLionProjects\pyo3\venv\lib\site-packages\pip\_vendor\pep517\_in_process.py", line 141, in build_wheel
      metadata_directory)
    File "C:\Users\azati\AppData\Local\Temp\pip-build-env-n1a5am83\overlay\Lib\site-packages\setuptools\build_meta.py", line 192, in build_wheel
      self.run_setup()
    File "C:\Users\azati\AppData\Local\Temp\pip-build-env-n1a5am83\overlay\Lib\site-packages\setuptools\build_meta.py", line 141, in run_setup
      exec(compile(code, __file__, 'exec'), locals())
    File "setup.py", line 92, in <module>
      'sdist': CargoModifiedSdist,
    File "C:\Users\azati\AppData\Local\Temp\pip-build-env-n1a5am83\overlay\Lib\site-packages\setuptools\__init__.py", line 145, in setup
      return distutils.core.setup(**attrs)
    File "C:\Python35\lib\distutils\core.py", line 148, in setup
      dist.run_commands()
    File "C:\Python35\lib\distutils\dist.py", line 955, in run_commands
      self.run_command(cmd)
    File "C:\Python35\lib\distutils\dist.py", line 974, in run_command
      cmd_obj.run()
    File "C:\Users\azati\AppData\Local\Temp\pip-build-env-n1a5am83\overlay\Lib\site-packages\wheel\bdist_wheel.py", line 192, in run
      self.run_command('build')
    File "C:\Python35\lib\distutils\cmd.py", line 313, in run_command
      self.distribution.run_command(command)
    File "C:\Python35\lib\distutils\dist.py", line 974, in run_command
      cmd_obj.run()
    File "C:\Python35\lib\distutils\command\build.py", line 135, in run
      self.run_command(cmd_name)
    File "C:\Python35\lib\distutils\cmd.py", line 313, in run_command
      self.distribution.run_command(command)
    File "C:\Python35\lib\distutils\dist.py", line 974, in run_command
      cmd_obj.run()
    File "C:\Users\azati\AppData\Local\Temp\pip-build-env-n1a5am83\overlay\Lib\site-packages\setuptools_rust\build_ext.py", line 26, in run
      build_rust.run()
    File "C:\Users\azati\AppData\Local\Temp\pip-build-env-n1a5am83\overlay\Lib\site-packages\setuptools_rust\build.py", line 313, in run
      self.build_extension(ext)
    File "C:\Users\azati\AppData\Local\Temp\pip-build-env-n1a5am83\overlay\Lib\site-packages\setuptools_rust\build.py", line 95, in build_extension
      metadata = json.loads(check_output(metadata_command).decode("utf-8"))
    File "C:\Python35\lib\subprocess.py", line 316, in check_output
      **kwargs).stdout
    File "C:\Python35\lib\subprocess.py", line 398, in run
      output=stdout, stderr=stderr)
  subprocess.CalledProcessError: Command '['cargo', 'metadata', '--manifest-path', 'Cargo.toml', '--format-version', '1']' returned non-zero exit status 101
  ----------------------------------------
  ERROR: Failed building wheel for word-count
  Running setup.py clean for word-count
Failed to build word-count
ERROR: Could not build wheels for word-count which use PEP 517 and cannot be installed directly
```
or it should be made differently?

lycantropos replied:
get similar output in WSL
```bash
(venv) $ python -m pip install -U pip
...
Successfully installed pip-19.1
(venv) $ python -m pip install .
Processing /home/ace/pyo3/examples/word-count
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
Building wheels for collected packages: word-count
  Building wheel for word-count (PEP 517) ... error
  ERROR: Complete output from command /home/ace/pyo3/examples/word-count/venv/bin/python /home/ace/pyo3/examples/word-count/venv/lib/python3.6/site-packages/pip/_vendor/pep517/_in_process.py build_wheel /tmp/tmp_ald7kvr:
  ERROR: running bdist_wheel
  running build
  running build_py
  running egg_info
  writing word_count.egg-info/PKG-INFO
  writing dependency_links to word_count.egg-info/dependency_links.txt
  writing top-level names to word_count.egg-info/top_level.txt
  reading manifest file 'word_count.egg-info/SOURCES.txt'
  reading manifest template 'MANIFEST.in'
  writing manifest file 'word_count.egg-info/SOURCES.txt'
  running build_ext
  running build_rust
  error: failed to load source for a dependency on `pyo3`

  Caused by:
    Unable to update /

  Caused by:
    failed to read `/Cargo.toml`

  Caused by:
    No such file or directory (os error 2)
  Traceback (most recent call last):
    File "/home/ace/pyo3/examples/word-count/venv/lib/python3.6/site-packages/pip/_vendor/pep517/_in_process.py", line 207, in <module>
      main()
    File "/home/ace/pyo3/examples/word-count/venv/lib/python3.6/site-packages/pip/_vendor/pep517/_in_process.py", line 197, in main
      json_out['return_val'] = hook(**hook_input['kwargs'])
    File "/home/ace/pyo3/examples/word-count/venv/lib/python3.6/site-packages/pip/_vendor/pep517/_in_process.py", line 141, in build_wheel
      metadata_directory)
    File "/tmp/pip-build-env-ir8gxftr/overlay/lib/python3.6/site-packages/setuptools/build_meta.py", line 192, in build_wheel
      self.run_setup()
    File "/tmp/pip-build-env-ir8gxftr/overlay/lib/python3.6/site-packages/setuptools/build_meta.py", line 141, in run_setup
      exec(compile(code, __file__, 'exec'), locals())
    File "setup.py", line 92, in <module>
      'sdist': CargoModifiedSdist,
    File "/tmp/pip-build-env-ir8gxftr/overlay/lib/python3.6/site-packages/setuptools/__init__.py", line 145, in setup
      return distutils.core.setup(**attrs)
    File "/usr/lib/python3.6/distutils/core.py", line 148, in setup
      dist.run_commands()
    File "/usr/lib/python3.6/distutils/dist.py", line 955, in run_commands
      self.run_command(cmd)
    File "/usr/lib/python3.6/distutils/dist.py", line 974, in run_command
      cmd_obj.run()
    File "/tmp/pip-build-env-ir8gxftr/overlay/lib/python3.6/site-packages/wheel/bdist_wheel.py", line 192, in run
      self.run_command('build')
    File "/usr/lib/python3.6/distutils/cmd.py", line 313, in run_command
      self.distribution.run_command(command)
    File "/usr/lib/python3.6/distutils/dist.py", line 974, in run_command
      cmd_obj.run()
    File "/usr/lib/python3.6/distutils/command/build.py", line 135, in run
      self.run_command(cmd_name)
    File "/usr/lib/python3.6/distutils/cmd.py", line 313, in run_command
      self.distribution.run_command(command)
    File "/usr/lib/python3.6/distutils/dist.py", line 974, in run_command
      cmd_obj.run()
    File "/tmp/pip-build-env-ir8gxftr/overlay/lib/python3.6/site-packages/setuptools_rust/build_ext.py", line 26, in run
      build_rust.run()
    File "/tmp/pip-build-env-ir8gxftr/overlay/lib/python3.6/site-packages/setuptools_rust/build.py", line 313, in run
      self.build_extension(ext)
    File "/tmp/pip-build-env-ir8gxftr/overlay/lib/python3.6/site-packages/setuptools_rust/build.py", line 95, in build_extension
      metadata = json.loads(check_output(metadata_command).decode("utf-8"))
    File "/usr/lib/python3.6/subprocess.py", line 336, in check_output
      **kwargs).stdout
    File "/usr/lib/python3.6/subprocess.py", line 418, in run
      output=stdout, stderr=stderr)
  subprocess.CalledProcessError: Command '['cargo', 'metadata', '--manifest-path', 'Cargo.toml', '--format-version', '1']' returned non-zero exit status 101.
  ----------------------------------------
  ERROR: Failed building wheel for word-count
  Running setup.py clean for word-count
Failed to build word-count
ERROR: Could not build wheels for word-count which use PEP 517 and cannot be installed directly
```

lycantropos commented:
> In addition to the issue I mentioned about `pip` or `pep517.build` vs `setup.py`, I think this is the wrong way to do this test. These are a bunch of system commands, so there's no need to write this in Rust, or to use conditional compilation here.
> 
> I think maybe start by just adding a script to the CI that you invoke on Appveyor that executes one of these builds.
> 
> I also don't see the _fix_ here, is that coming in a later commit?

There should be different script for Travis CI as well? To be sure that I'm not breaking something. If so, where to place scripts? Directly in YAML configurations?

kngwyu commented:
> I think maybe start by just adding a script to the CI that you invoke on Appveyor that executes one of these builds.

:+1:

> There should be different script for Travis CI as well? To be sure that I'm not breaking something. If so, where to place scripts? Directly in YAML configurations?

In ci/travis/test.sh or so.

pganssle commented:
@lycantropos @kngwyu I think Appveyor is the important one, but having it on Travis doesn't hurt.

I think maybe `ci/travis` is the wrong place, maybe just `ci/test_venv.sh` or something like that.

kngwyu commented:
How about using `tox` to test venv?
Then we don't need to add additional scripts.

lycantropos commented:
what should we do with [this kind of message](https://travis-ci.org/PyO3/pyo3/jobs/528387287#L353):
```bash
./ci/travis/test.sh: line 6: cargo: command not found
The command "./ci/travis/test.sh" exited with 127.
```
?
it seems like modifying `PATH` in *setup.sh* is not working

lycantropos commented:
> How about using `tox` to test venv?
> Then we don't need to add additional scripts.

I'm not so familiar with `tox`, can you please provide an example of activating `venv` in it?

konstin commented on .travis.yml:
```diff
@@ -46,13 +46,14 @@ addons:
       - libiberty-dev
 
 before_install:
-  - source ./ci/travis/setup.sh
```
This change is the reason for 

```
./ci/travis/test.sh: line 6: cargo: command not found
```

konstin commented:
>  I'm not so familiar with tox, can you please provide an example of activating venv in it?

Not kngwyu, but anyways: The cool thing about tox is that it creates the venv itself. If the correct python version is active, you should only need to call `tox -e py` (and maybe `pip install tox` beforehand). You can easily try this out by just running `tox` on your local machine

lycantropos commented:
@konstin: so in the end it will be similar to *ci/travis* directory (especially *setup.sh* & *test.sh* scripts), but for AppVeyor?

pganssle commented:
> How about using tox to test venv?
> Then we don't need to add additional scripts.

I'm not entirely sure why our existing tests with `tox` were not hitting this edge case already, honestly, which is why I didn't suggest `tox` in the first place.

kngwyu commented:
@pganssle 
I think now tox is not invoked in Appveyor.
Or you tested the tox works on local windows?

pganssle commented:
@kngwyu Oh, I assumed the tests were run on all platforms. Can we just turn on tox for the appveyor builds and have that test this?

I have no Windows computer to test this on, so I haven't done anything local.

kngwyu commented:
@pganssle 
> Can we just turn on tox for the appveyor builds and have that test this?

I tried it in #475

pganssle commented:
@kngwyu Looks like that successfully reproduces the issue!

@lycantropos Sorry I didn't suggest this earlier, I genuinely thought we were already doing it! I think enabling the `tox` runs on appveyor like @kngwyu did in #475 is the best way to test your change.

I also checked with @gaborbernat, who maintains virtualenv, and his cursory checks seemed to indicate that `base_prefix` is the right thing to use, so your approach to fixing this looks good.

lycantropos commented:
I've added `tox-venv` plugin to use stdlib `venv` instead of `virtualenv` because the latter seems to create full copy of Python and overriding `sys.base_prefix` as well, but didn't make a copy of *Python/libs* directory which is needed for projects to compile on Windows.
I've also added `pip` & `setuptools` in examples' requirements because it seems like they need to be upgraded manually until [this PR](https://github.com/tox-dev/tox-venv/pull/25) is merged and `test-venv` is updated.

Finally, we have errors related to `hypothesis` strategies used for [`datetime.timestamp` method](https://docs.python.org/3/library/datetime.html#datetime.datetime.timestamp) testing, which min/max limits are [platform-dependent](https://stackoverflow.com/a/50860754/5997596).

lycantropos commented:
I've manually found edge-cases for `datetime.timestamp`, for Linux environment I've used *Docker* containers.

lycantropos commented on .travis.yml:
```diff
@@ -49,7 +49,7 @@ before_install:
+  - pip install setuptools-rust pytest pytest-benchmark tox tox-venv
 
 script:
   - ./ci/travis/test.sh
```
if I understood correctly we can add `virtualenv` support for Linux by installing `tox-venv` and running  script as a separate step like
```yml
install:
  - pip install setuptools-rust pytest pytest-benchmark tox

script:
  - ./ci/travis/test.sh
  - pip install tox-venv
  - ./ci/travis/test.sh
```
but it will increase build time

lycantropos commented on ci/appveyor/setup.ps1:
```diff
@@ -0,0 +1,8 @@
+$env:PATH="$env:PATH;$((Get-Location).Path)/rust-nightly-$env:TARGET/bin"
+
+$pythonLocation = Invoke-Expression "python -c `"import sys; print(sys.base_prefix)`""
+$env:LIBPATH = "$env:LIBPATH; $( Join-Path $pythonLocation "libs" )"
```
it is similar to this line
https://github.com/PyO3/pyo3/blob/c4c75bbf811773414d31ae9205988f47ee876399/ci/travis/setup.sh#L35

lycantropos commented on appveyor.yml:
```diff
@@ -1,25 +1,24 @@
+  RUST_BACKTRACE: 1
   matrix:
     - PYTHON: "C:/Python35-x64"
     - PYTHON: "C:/Python36-x64"
```
what about adding `x32` versions and Python37? this can be done in a separate PR though

kngwyu replied:
Yeah, I think we should add Python37 but I'm not sure we should support 32bit architecture.
And just a nit, but there's no architecture named 'x32'.
The 32bit version of 'x86-64' is called 'x86'.

lycantropos replied:
Thanks, I'll keep it in mind.

kngwyu commented:
@lycantropos 
Good catch, thanks.
But probably the fix for the datetime example should be in a separate PR.

lycantropos commented:
> @lycantropos
> Good catch, thanks.
> But probably the fix for the datetime example should be in a separate PR.

I thought that there is no way of merging if tests fail. Also I do not know what is left to do or I can add entry to changelog and we are done?

pganssle commented on examples/word-count/pyproject.toml:
```diff
@@ -1,3 +1,3 @@
 [build-system]
-requires = ["setuptools", "wheel", "setuptools-rust", "toml"]
+requires = ["setuptools>=41.0.0", "pip>=19.1", "wheel", "setuptools_rust>=0.10.2", "toml"]
```
That doesn't make sense to me, `pip` is invoked *before* this file is read. I don't think it will have any effect on the build.

pganssle commented on examples/word-count/requirements-dev.txt:
```diff
@@ -1,3 +1,5 @@
+setuptools>=41.0.0
+pip>=19.1
```
I think you may need `pip>=19.1` here, but not `setuptools`, which will not be used.

lycantropos replied:
you are right, completed dependencies

lycantropos commented on examples/rustapi_module/requirements-dev.txt:
```diff
@@ -1,3 +1,5 @@
+setuptools>=41.0.0
+pip>=19.1
```
left `pip`, it fails to build without it

kngwyu commented:
I changed the title instead of separating the PR

kngwyu commented on build.rs:
```diff
@@ -534,7 +534,7 @@ fn configure(interpreter_version: &PythonVersion, lines: Vec<String>) -> Result<
     let enable_shared: &str = &lines[2];
     let ld_version: &str = &lines[3];
-    let exec_prefix: &str = &lines[4];
+    let base_prefix: &str = &lines[4];
```
This type annotation(`: &str`) is redundant.
Could you please all `: &str` from these 4 lines?
They're not your fault, but it's nice timing to refactor them.

kngwyu commented:
It's merged at last, thanks.
